compiler/rustc_arena/src/lib.rs

```diff
@@ -167,6 +198,25 @@ impl<T> IterExt<T> for Vec<T> {
     }
 }
 
+impl<'a, T: 'a> IterExt<T> for Cloned<Iter<'a, T>>
+where
+    T: Clonable,
+{
+    #[inline]
+    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
+        let len = self.len();
+        if len == 0 {
+            return &mut [];
+        }
+        // Move the content to the arena by copying and then forgetting it
+        unsafe {
+            let len = self.len();
+            let start_ptr = arena.alloc_raw_slice(len);
+            write_from_iter(self, len, start_ptr)
+        }
+    }
```

Comment:
This doesn't seem to rely on anything specific to `Cloned`. It just relies on the length. Wouldn't specializing on `TrustedLen` be sufficient here, considering that's already a specialization trait?

Reply:
No, it wouldn't compile unfortunately, or at least I couldn't get it to. Cloned needs Clone on T.

Reply:
Also, just specializing on anything TrustedLen is unsound, see #77945 for that. In short, Map is TrustedLen and can allocate on the same arena, which makes handling a panic safely an impossible nightmare of dropping uninitialized memory.

I went through the uses of alloc_from_iter and I found that most of those are Map and FlatMap, but some places allocate from Cloned<slice::Iter> so hence the PR.

Reply:
I meant `impl<I: TrustedLen> IterExt for I`. That'll give you the length.

Reply:
Well it's closed already. But for the record, you might have been able to use use `impl<I: TrustedLen + IsClonedSliceIterMarker> for I` instead where the latter bound is a safe specialization marker.

Reply:
Indeed, this seems to compile:
```rust
#[rustc_specialization_trait]
trait IsClonedSliceIterMarker {}
impl<T> IsClonedSliceIterMarker for Cloned<Iter<'_, T>> {}

impl<T, I> IterExt<T> for I
where
    I: IntoIterator<Item = T> + IsClonedSliceIterMarker + TrustedLen,
{
    #[inline]
    fn alloc_from_iter(self, arena: &TypedArena<T>) -> &mut [T] {
        let iter = self.into_iter();
        let len = iter.size_hint().0;
        if len == 0 {
            return &mut [];
        }
        // Move the content to the arena by copying and then forgetting it
        unsafe {
            let start_ptr = arena.alloc_raw_slice(len);
            write_from_iter(iter, len, start_ptr)
        }
    }
}
```